Mathlib/Analysis/Calculus/Gradient/Calculation.lean

```diff
@@ -0,0 +1,400 @@
+import Mathlib.Analysis.Calculus.Gradient.Basic
+
+/-!
+# Gradient
+
+## Main results
+
+This file contains the following parts of gradient.
+* the chain rule for the `g : 𝕜 → 𝕜` composed with `f : F → 𝕜`.
+* the gradient for the sum of two functions.
+* the gradient for finite sum of functions.
+* the gradient for the sum of a constant and a function.
+* the gradient for the difference of two functions.
+* the gradient for the difference of a constant and a function.
+* the gradient for the product of two functions.
+* the gradient for the product of a constant and a function.
+-/
+
+noncomputable section
+
+open Topology InnerProductSpace Set
+
+variable {𝕜 F : Type*} [IsROrC 𝕜]
+
+variable [NormedAddCommGroup F] [InnerProductSpace 𝕜 F] [CompleteSpace F]
+
+variable {f : F → 𝕜} {f' x : F} {L : Filter F} {s : Set F}
+
+local notation "∇" => gradient
+
+section Composition
+
+open Set Filter
+
+variable {g : 𝕜 → 𝕜} {g' : 𝕜}
+
+variable {L' : Filter 𝕜} {t : Set 𝕜}
+
+theorem HasGradientAtFilter.comp
+    (hg : HasGradientAtFilter g g' (f x) L') (hf : HasGradientAtFilter f f' x L)
+    (hL : Tendsto f L L') : HasGradientAtFilter (g ∘ f) (g' • f') x L := by
+  have eq : (starRingEnd 𝕜) g' • (toDual 𝕜 F) f' =
+    (toDual 𝕜 F) (g' • f') := by rw [map_smulₛₗ]
+  rw [HasGradientAtFilter, ← eq]
+  exact hg.hasDerivAtFilter.comp_hasFDerivAtFilter x hf hL
+
+theorem HasGradientWithinAt.comp
+    (hg : HasGradientWithinAt g g' t (f x)) (hf : HasGradientWithinAt f f' s x)
+    (hst : MapsTo f s t) : HasGradientWithinAt (g ∘ f) (g' • f') s x :=
+  HasGradientAtFilter.comp hg hf <| hf.continuousWithinAt.tendsto_nhdsWithin hst
+
+theorem HasGradientAt.comp_hasGradientWithinAt
+    (hg : HasGradientAt g g' (f x)) (hf : HasGradientWithinAt f f' s x) :
+    HasGradientWithinAt (g ∘ f) (g' • f') s x :=
+  hg.comp hf hf.continuousWithinAt
+
+theorem HasGradientWithinAt.comp_of_mem
+    (hg : HasGradientWithinAt g g' t (f x)) (hf : HasGradientWithinAt f f' s x)
+    (hst : Tendsto f (𝓝[s] x) (𝓝[t] f x)) : HasGradientWithinAt (g ∘ f) (g' • f') s x :=
+  HasGradientAtFilter.comp hg hf hst
+
+/-- The chain rule. -/
+theorem HasGradientAt.comp (hg : HasGradientAt g g' (f x))
+    (hf : HasGradientAt f f' x) : HasGradientAt (g ∘ f) (g' • f') x :=
+  HasGradientAtFilter.comp hg hf hf.continuousAt
+
+theorem gradient.comp (hg : DifferentiableAt 𝕜 g (f x)) (hf : DifferentiableAt 𝕜 f x) :
+    ∇ (g ∘ f) x = (∇ g (f x)) • (∇ f x) :=
+  (hg.hasGradientAt.comp hf.hasGradientAt).gradient
+
+end Composition
+
+section ConstSmul
+
+open Set Filter
+
+theorem HasGradientAtFilter.const_smul (h : HasGradientAtFilter f f' x L) (c : 𝕜) :
+    HasGradientAtFilter (fun x => c • f x) ((starRingEnd 𝕜) c • f') x L := by
+  have : c • (toDual 𝕜 F) f' = (toDual 𝕜 F) ((starRingEnd 𝕜) c • f') := by
+    rw [map_smulₛₗ, RingHomCompTriple.comp_apply, RingHom.id_apply]
+  rw [HasGradientAtFilter, ← this]; rw [HasGradientAtFilter] at h
+  exact h.const_smul c
```

Comment:
The preferred style is to write each tactic on a separate line, instead of using `;` on the same line. That's superfluous anyway because that line is not used.
```suggestion
  rw [HasGradientAtFilter, map_smulₛₗ, RingHomCompTriple.comp_apply, RingHom.id_apply]
  exact HasFDerivAtFilter.const_smul h c
```